internal/operands/common-templates/reconcile.go

```diff
@@ -167,58 +227,68 @@ func getOldTemplatesLabelSelector() labels.Selector {
 	return labels.NewSelector().Add(*baseRequirement, *versionRequirement)
 }
 
-func (c *commonTemplates) reconcileOlderTemplates(request *common.Request) ([]common.ReconcileFunc, error) {
-	existingTemplates := &templatev1.TemplateList{}
-	err := request.Client.List(request.Context, existingTemplates, &client.ListOptions{
+func (c *commonTemplates) deprecateOrDeleteOldTemplates(request *common.Request, deployedTemplates []templatev1.Template, archs []architecture.Arch) ([]common.ReconcileFunc, error) {
+	oldTemplates := &templatev1.TemplateList{}
+	err := request.Client.List(request.Context, oldTemplates, &client.ListOptions{
 		LabelSelector: getOldTemplatesLabelSelector(),
 		Namespace:     request.Instance.Spec.CommonTemplates.Namespace,
 	})
-
 	// There might not be any templates (in case of a fresh deployment), so a NotFound error is accepted
 	if err != nil && !errors.IsNotFound(err) {
 		return nil, err
 	}
 
-	templatesVersion, err := semver.ParseTolerant(Version)
+	latestTemplateVersion, err := semver.ParseTolerant(Version)
 	if err != nil {
 		return nil, err
 	}
 
-	funcs := make([]common.ReconcileFunc, 0, len(existingTemplates.Items))
-	for i := range existingTemplates.Items {
-		template := &existingTemplates.Items[i]
+	ownedTemplates, err := common.ListOwnedResources[templatev1.TemplateList, templatev1.Template](request,
+		client.InNamespace(request.Instance.Spec.CommonTemplates.Namespace))
+	if err != nil {
+		return nil, fmt.Errorf("failed to list owned templates: %w", err)
+	}
+
+	nonDeployedTemplates := map[string]templatev1.Template{}
+	for _, template := range oldTemplates.Items {
+		nonDeployedTemplates[template.Name] = template
+	}
+	for _, template := range ownedTemplates {
+		nonDeployedTemplates[template.Name] = template
+	}
+	for i := range deployedTemplates {
+		delete(nonDeployedTemplates, deployedTemplates[i].Name)
+	}
 
-		if _, ok := c.deployedTemplates[template.Name]; ok {
+	var funcs []common.ReconcileFunc
+	for _, template := range nonDeployedTemplates {
+		if !template.DeletionTimestamp.IsZero() {
 			continue
 		}
 
-		// if template has higher version than is defined in ssp operator, keep it as it is. If parsing
-		// of template version fails, continue with adding deprecated label
+		// Delete the template if it is not one of the cluster architectures
+		templateArch, err := GetTemplateArch(&template)
+		if err != nil {
+			// On error, we assume that the template has unknown architecture,
+			// and it will be deleted.
+			templateArch = ""
+		}
+
+		if !slices.Contains(archs, templateArch) {
+			funcs = append(funcs, reconcileDeleteTemplate(&template))
+			continue
+		}
```

Comment:
Why are these checks necessary? Templates will be deleted regardless once L291 is reached?

Reply:
This function handles two cases for existing templates, that are owned by SSP, but are not currently being deployed (they are not in `deployedTemplates` parameter).

- If the existing template has architecture that is no longer supported by the cluster, the template is removed.
- If the existing template has architecture that is supported, the template is deprecated, by adding annotation and label.

Reply:
To keep the same backward compatible behavior. So far, before this PR, when a common template was removed from SSP, for example an old RHEL, we would keep the template in the cluster, because they may still be using it.

But now with multi-arch, when all nodes of a specific architecture are removed from the cluster, it does not make sense to keep the old template, because it cannot be run.